shell/platform/android/io/flutter/embedding/engine/FlutterEngine.java

```diff
@@ -289,6 +289,7 @@ public void destroy() {
     Log.v(TAG, "Destroying.");
     // The order that these things are destroyed is important.
     pluginRegistry.destroy();
+    platformViewsController.onFlutterViewDestroyed();
```

Comment:
I'm not sure this actually belongs here. Notice that `PlatformViewsController` is injected in the constructor. As such, `FlutterEngine` is not the de facto owner of the `PlatformViewsController`. Instead, the object that provided the `PlatformViewsControllers` owns it and should take care to initialize and destroy it when desired. The reason this is important is because `FlutterEngine` does not know the intentions of the object that owns it. Perhaps the `PlatformViewsController` needs to outlive this `FlutterEngine`. That situation is made impossible by forcibly destroying its `View` relationship here.

Reply:
I'm also seeing that we have a constructor that automatically creates one. Perhaps we need some mechanism for tracking whether it was legitimately injected vs created by `FlutterEngine`?

Reply:
The `PlatformViewsController` holds references to virtual display controllers and their textures.  When those textures are updated, the Flutter engine is notified about the update.  That notification is no longer valid once the `FlutterEngine` is detached from its native peer.

Calling `PlatformViewsController.onFlutterViewDestroyed` is intended to signal that these notifications should stop.  Is there a better way to do this?

Reply:
I think the goal should some kind of symmetry, in some fashion. So if it makes sense that `FlutterEngine` invokes `onFlutterViewDestroyed()` then I think it would make sense that `PlatformViewsController` does not "initialize" until `FlutterEngine` tells it to do so. This way, when the `PlatformViewsController` comes into `FlutterEngine`, `FlutterEngine` can inform it `onFlutterViewAttached()` or some such message, and then the `onFlutterViewDestroyed()` would make sense in the place where you added it. 

Does that symmetry notion make sense?

I haven't had a chance to dig into what `PlatformViewsController` is doing, but given that `FlutterEngine` is independent from `FlutterView`, do you know why the concept of the `FlutterView` being destroyed is a message that is expected to come from `FlutterEngine`?

Reply:
Could this be solved by renaming `PlatformViewsController.onFlutterViewDestroyed` to `onFlutterEngineDestroyed`?

The name `onFlutterViewDestroyed` is a legacy from `io.flutter.view.FlutterView`/`FlutterNativeView`.  `FlutterNativeView.destroy` would call `FlutterPluginRegistry.destroy`/`PlatformViewsController.destroy` before it detached from the JNI peer.

`FlutterEngine` replaces `FlutterNativeView` in the new embedding, and `FlutterEngine` needs to make a similar notification to the `PlatformViewsController`.  But I think it would make sense to change the name of the method to reflect that the lifecycle of the textures managed by `PlatformViewsController` is tied to the engine lifecycle and not to the view.

Reply:
Seems like for all intents and purposes, the FlutterEngine owns the PlatformViewController since I don't see the PlatformViewController injection happening anywhere in our codebase or in g3 except for this test https://github.com/flutter/engine/blob/master/shell/platform/android/test/io/flutter/embedding/engine/FlutterEngineTest.java#L71. Can we annotate that engine constructor `@VisibleForTesting` to avoid this confusion? 

Can we mark onFlutterViewDestroyed deprecated and create a new onFlutterEngineDestroyed() or maybe just dispose()?

Reply:
@jason-simmons that rename seems to make sense to me. I would still suggest creating symmetry. I think anything that happens in an object's constructor should only be undone by object deallocation, and anything done to an object in a method, should likewise be undone in another method.

So if you introduce `onFlutterEngineDestroyed()` then I would still recommend an additional `onFlutterEngineAvailable(engine)` or `onFlutterEngineCreated(engine)` or some such symmetric method that is invoked early in `PlatformViewController`'s lifetime.